0012-unicode-strings.md

```diff
@@ -0,0 +1,377 @@
+and instead uses the 32-bit wide `int` primitive type.
+
+## Go:
+
+In Go a string is a read-only slice of bytes, which generally contains UTF-8 encoded bytes
+but may contain arbitrary bytes. The `rune` type is an alias for `int32`.
+
+## JavaScript:
+
+The JavaScript `String` type represents an immutable sequence of UTF-16 code units as integer values.
+There is no distinct type for representing individual characters.
+
+## C++:
+
+The `char` type represents bytes, and the `std::string` class from the standard library
+operates on bytes as character, and generally does not produce correct results if used
+together with any encoding other than single-byte encodings such as ASCII.
+C++11 added two new character types, `char16_t` and `char32_t`, 
+and two new corresponding `std::u16string` and `std::u32string` collection classes.
+It also provides three new kinds of string literals,
+`u8"..."`, `u"..."`, and `U"..."`,
+which produce binary values encoded with UTF-8, UTF-16, and UTF-32 respectively.
+
+## Python:
+
+In Python 2, the `str` type was used for both binary and text data.
+A separate `unicode` type was introduced to store textual data,
+and a separate mode for unicode string literals (`u"..."`).
+Python 3 changed `str` to be this `unicode` datatype instead.
+Like JavaScript, Python has no dedicated character type,
+and just uses the unbounded `int` type for Unicode scalar values.
+
+## Rust:
+
+The `char` type directly represents any Unicode scalar value, and rejects out-of-range values.
+The "string slice" type `str` is the most primitive string type,
+which is guaranteed to refers to valid UTF-8 data.
+It primarily acts as a sequence of bytes, but directly provides many methods that work with `char` values:
+`chars()` provides an iterator of `char` values,
+`get(<slice>)` returns a subslice but produces `None` if the bounds do not align with UTF-8 sequence boundaries,
+etc.
+
+## Swift:
+
+Swift defines `String` as a collection of characters,
+and unusually defines `Character` as an [*extended grapheme cluster*](https://unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries),
+which is the closest concept to an atomic characeter as perceived visually by users.
+Unicode scalar values are represented with a distinct `Unicode.Scalar` structure instead.
+`String` provides `utf8`, `utf16`, and `unicodeScalars` views
+that present the contents as sequences of UTF-8, UTF-16, and UTF-32 code units respectively.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Is there anything more we can do to make migration easier and safer for users?
+Chance are very good that all Dafny code in existence to date either will not change behavior
+across this change, or will slightly improve because of the improved handling of surrogate code points.
+I have been unable to think of anything that would provide more value than the verifier will already provide,
+but I am open to suggestion as always!
```

Comment:
Automatically converting old-style `\u` strings to new-style `\u{}` strings would be useful :)

Reply:
That's a fair point, and although I've clarified I don't intend to drop `\u` (just reject invalid sequences of them), even so that conversion utility would be very easy to implement.

Reply:
Now that I'm dropping `\u` I agree and will definitely provide a conversion utility.